### PR TITLE
feat: add udf sparrow trait 

### PR DESCRIPTION
**Pull request by jordanrfrazier**
Adds the sparrow trait for user defined functions. This will allow us to both keep python dependencies in `python` crates and implement evaluators from other `crates`.

### CODE REVIEW

**Comment by bjchambers** [comment]
> Adds the sparrow trait for user defined functions. This will allow us to keep python dependencies in python crates, while passing

While passing *what*?

File: crates/sparrow-instructions/src/inst.rs [hunk 1]
```diff
@@ -236,15 +232,43 @@ pub enum InstKind {
...
+            Self::Simple(arg0) => Self::Simple(arg0.clone()),
+            Self::FieldRef => Self::FieldRef,
+            Self::Cast(arg0) => Self::Cast(arg0.clone()),
+            Self::Record => Self::Record,
+            Self::Udf(arg0) => Self::Udf(arg0.clone()),
+        }
+    }
+}
+
+impl PartialEq for InstKind {
```
Comment (jordanrfrazier) [inline crates/sparrow-instructions/src/inst.rs:251]: I can't derive `PartialEq` -- something about it moving `self` out. Can look at this together if you think we can get around that.

File: crates/sparrow-instructions/src/inst.rs [hunk 2]
```diff
@@ -236,15 +232,43 @@ pub enum InstKind {
...
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::Simple(l0), Self::Simple(r0)) => l0 == r0,
+            (Self::FieldRef, Self::FieldRef) => true,
+            (Self::Cast(l0), Self::Cast(r0)) => l0 == r0,
+            (Self::Record, Self::Record) => true,
+            (Self::Udf(l0), Self::Udf(r0)) => l0.eq(r0),
```
Comment (bjchambers) [inline crates/sparrow-instructions/src/inst.rs:258]: I think we should be able to write this `l0 == r0`

File: crates/sparrow-instructions/src/inst.rs (hunk 1 above)
Comment (bjchambers) [reply crates/sparrow-instructions/src/inst.rs:251]: that's somewhat surprising -- I can go either way.

**Review by bjchambers — APPROVED** [review]
Some clippy errors you should fix.

**Pull request merged**